### ISSUE TITLE
[3.1.2+] this._routerMicrolib is undefined when sending an event (in tests only?)

### ISSUE THREAD

**Issue by backspace** [emberjs/ember.js#16633]
Hey, I’ve been looking at [updating `travis-web` to Ember 3.1](https://github.com/travis-ci/travis-web/pull/1665); everything seems to be working except some failures in tests around dynamically updating the favicon based on job/build status. I’ve tried it with Ember 3.1.2 and 3.2.0-beta.4.

When running the tests, many of them [fail](https://travis-ci.org/travis-ci/travis-web/builds/378750218#L639) with this message (our suite uses Chrome):

> Cannot read property 'trigger' of undefined

In digging into the stacktrace, I found that the [line in question](https://github.com/emberjs/ember.js/blob/470dddb9bbea32f0bf71219740b22c8dc8ecf68d/packages/ember-routing/lib/system/router.js#L440) is in `packages/ember-routing/lib/system/router.js`:

```
  send() {
    /*name, context*/
    this._routerMicrolib.trigger(...arguments);
  },
```

That’s [called by](https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/mixins/action_handler.js#L199-L219) `packages/ember-runtime/lib/mixins/action_handler.js#send`.

Lower in the stack, there’s some `travis-web`-specific code:

```
TypeError: Cannot read property 'trigger' of undefined
    at Class.send (http://localhost:4200/assets/vendor.js:56252:49)
    at Class.send (http://localhost:4200/assets/vendor.js:60528:21)
    at Class.buildStateDidChange (http://localhost:4200/assets/travis.js:8780:14)
```

The job controller has [an observer](https://github.com/travis-ci/travis-web/blob/873cc0e63642f0e878fc6f4dba50739975e86da2/app/controllers/job.js#L21-L24) for `job.state`:

```
  @observes('job.state')
  jobStateDidChange() {
    return this.send('faviconStateDidChange', this.get('job.state'));
  },
```

The build controller has something [almost identical](https://github.com/travis-ci/travis-web/blob/873cc0e63642f0e878fc6f4dba50739975e86da2/app/controllers/build.js#L36-L41).

The `faviconStateDidChange` action is [defined](https://github.com/travis-ci/travis-web/blob/873cc0e63642f0e878fc6f4dba50739975e86da2/app/mixins/build-favicon.js#L8-L14) in a `build-favicon` mixin that’s _only_ [mixed into the application route](https://github.com/travis-ci/travis-web/blob/cf651111b9181867f2a56e2c14ace016df3413e4/app/routes/application.js#L11):

```
    faviconStateDidChange(state) {
      if (state) {
        return this.setFaviconForState(state);
      } else {
        return this.setDefault();
      }
    }
```

This all looks innocuous to me; I tried removing the code within the action and that didn’t change anything. I also tried merging the mixin’s code directly into the application route, which also didn’t do anything. It seems like the code to bubble the action is the problem more than the action code itself.

A strange aspect of this is that when I actually use the interface myself rather than run the tests, I can see that the favicon is updating, so the action is indeed being sent and run. So it seems like it’s only a problem in tests? 🤔 

Let me know if there’s any other information I can provide! Thanks for all your work on Ember, I’m a multi-year devotee 😀

**Comment by rwjblue** [comment]
@backspace - Can you link me to one of the failing tests in travis-web?

**Comment by backspace** [comment]
I tried updating to 3.2.0 this morning and while we still have 25 failing tests, that’s far fewer than the 100+ that were failing for what I described above and none of the failures seem similar to this one, so it looks like this has been fixed! 😀

**Comment by mweidner037** [comment]
I encountered this error in an integration test on a component that called `router.transitionTo`. We previously worked around it by using `this.owner.register("router:main", MyRouterStub)` to replace the router service with a stub, but that stopped working when we upgraded something unrelated (apparently [service stubs are brittle](https://github.com/ember-cli/ember-cli-qunit/issues/203#issuecomment-781550153)). My new workaround is to replace the stub with
```ts
this.owner.lookup("router:main").setupRouter()
```
in our `beforeEach` hook, following [this post](https://discuss.emberjs.com/t/to-stub-or-not-to-stub-the-router-service-in-integration-tests/16742/9).
